diff --git a/src/Departures Board.cpp b/src/Departures Board.cpp
index ea6452a..3ba3837 100644
--- a/src/Departures Board.cpp	
+++ b/src/Departures Board.cpp	
@@ -16,23 +16,33 @@
  * 14 D/C#        IO5
  * 16 CS#         IO26
  *
+ * Optional TTP223 touch sensor connection:
+ *
+ * TTP223         ESP32 MINI
+ * GND            GND
+ * VCC            3.3V
+ * I/O            IO34
+ *
  */
 
 // Release version number
 #define VERSION_MAJOR 2
-#define VERSION_MINOR 2
+#define VERSION_MINOR 3
 
 #include <Arduino.h>
 #include <WiFi.h>
 #include <WiFiClientSecure.h>
-#include <WebServer.h>
+#include <WiFiManager.h>
+#include <AsyncTCP.h>
+#include <ESPAsyncWebServer.h>
+#include <StreamString.h>
+#include <Ticker.h>
 #include <HTTPClient.h>
 #include <ESPmDNS.h>
 #include <HTTPUpdateGitHub.h>
 #include <FS.h>
 #include <LittleFS.h>
 #include <ArduinoJson.h>
-#include <WiFiManager.h>
 #include <weatherClient.h>
 #include <stationData.h>
 #include <raildataXmlClient.h>
@@ -40,9 +50,11 @@
 #include <busDataClient.h>
 #include <githubClient.h>
 #include <rssClient.h>
+#include <touchSensor.h>
 #include <webgui/webgraphics.h>
 #include <webgui/index.h>
 #include <webgui/keys.h>
+#include <webgui/editrss.h>
 #include <webgui/rssfeeds.h>
 #include <gfx/xbmgfx.h>
 
@@ -55,8 +67,7 @@
 #define msHour 3600000 // 3600000 milliseconds in an hour
 #define msMin 60000 // 60000 milliseconds in a second
 
-WebServer server(80);     // Hosting the Web GUI
-File fsUploadFile;        // File uploads
+static AsyncWebServer server(80); // Hosting the Web GUI
 
 // Shorthand for response formats
 static const char contentTypeJson[] PROGMEM = "application/json";
@@ -143,6 +154,8 @@ U8G2_SSD1322_NHD_256X64_F_4W_HW_SPI u8g2(U8G2_R0, /* cs=*/ 26, /* dc=*/ 5, /* re
 #define ULINE3 41
 #define ULINE4 56
 
+static Ticker restartTimer; // used to schedule reboots
+
 //
 // Custom fonts - replicas of those used on the real display boards
 //
@@ -277,16 +290,16 @@ static const uint8_t Underground10[1085] U8G2_FONT_SECTION("Underground10") =
   "\16N\204\27I\230)Q\232%Y\322\200\5\0\204\20\201\6\33\237\37\17\0\0\0";
 
 static const uint8_t UndergroundClock8[150] U8G2_FONT_SECTION("UndergroundClock8") =
-"\13\0\3\3\3\4\2\2\5\7\10\0\0\10\0\10\0\0\0\0\0\0}\60\12G\305\251\310\370&\251"
-"\0\61\10\304\305\222\234\364\0\62\15G\305\251\310\244B\331L(;\10\63\15G\305\251\310\244\42\62\215"
-"&\251\0\64\15G\305\24\316H\22%\311Q*\1\65\14G\305\70\310\250f\32MR\1\66\14G\305"
-"\251\310\250\26\31\233\244\2\67\12G\305\70\310\204z\225\2\70\15G\305\251\310h\222\212\214MR\1\71"
-"\15G\305\251\310h\22+\215&\251\0:\6\262\257 \22\0\0\0";
+  "\13\0\3\3\3\4\2\2\5\7\10\0\0\10\0\10\0\0\0\0\0\0}\60\12G\305\251\310\370&\251"
+  "\0\61\10\304\305\222\234\364\0\62\15G\305\251\310\244B\331L(;\10\63\15G\305\251\310\244\42\62\215"
+  "&\251\0\64\15G\305\24\316H\22%\311Q*\1\65\14G\305\70\310\250f\32MR\1\66\14G\305"
+  "\251\310\250\26\31\233\244\2\67\12G\305\70\310\204z\225\2\70\15G\305\251\310h\222\212\214MR\1\71"
+  "\15G\305\251\310h\22+\215&\251\0:\6\262\257 \22\0\0\0";
 
 // Service attribution texts
-const char nrAttributionn[] = "Powered by National Rail Enquiries";
-const char tflAttribution[] = "Powered by TfL Open Data";
-const char btAttribution[] = "Powered by bustimes.org";
+static const char nrAttributionn[] = "Powered by National Rail Enquiries";
+static const char tflAttribution[] = "Powered by TfL Open Data";
+static const char btAttribution[] = "Powered by bustimes.org";
 
 //
 // GitHub Client for firmware updates
@@ -301,77 +314,84 @@ github ghUpdate("","");
 #define BUSDATAUPDATEINTERVAL 45000   // How often we fetch data from bustimes.org (ms - 45 secs)
 
 // Bit and bobs
-unsigned long timer = 0;
-bool isSleeping = false;            // Is the screen sleeping (showing the "screensaver")
-bool sleepEnabled = false;          // Is overnight sleep enabled?
-bool forcedSleep = false;           // Is the system in manual sleep mode?
-bool sleepClock = true;             // Showing the clock in sleep mode?
-bool dateEnabled = false;           // Showing the date on screen?
-bool weatherEnabled = false;        // Showing weather at station location. Requires an OpenWeatherMap API key.
-bool enableBus = false;             // Include Bus services on the board?
-bool firmwareUpdates = true;        // Check for and install firmware updates automatically at boot?
-bool dailyUpdateCheck = false;      // Check for and install firmware updates at midnight?
-byte sleepStarts = 0;               // Hour at which the overnight sleep (screensaver) begins
-byte sleepEnds = 6;                 // Hour at which the overnight sleep (screensaver) ends
-int brightness = 50;                // Initial brightness level of the OLED screen
-unsigned long lastWiFiReconnect=0;  // Last WiFi reconnection time (millis)
-bool firstLoad = true;              // Are we loading for the first time (no station config)?
-int prevProgressBarPosition=0;      // Used for progress bar smooth animation
-int startupProgressPercent;         // Initialisation progress
-bool wifiConnected = false;         // Connected to WiFi?
-unsigned long nextDataUpdate = 0;   // Next National Rail update time (millis)
-int dataLoadSuccess = 0;            // Count of successful data downloads
-int dataLoadFailure = 0;            // Count of failed data downloads
-unsigned long lastLoadFailure = 0;  // When the last failure occurred
-int dateWidth;                      // Width of the displayed date in pixels
-int dateDay;                        // Day of the month of displayed date
-bool altStationEnabled = false;     // Switch between stations based on time of day
-bool altStationActive = false;      // Is the alternate station currently shown
-byte altStarts = 12;                // Hour at which to switch to the alternate station
-byte altEnds = 23;                  // Hour at which to switch back to the default station
-bool noScrolling = false;           // Suppress all horizontal scrolling
-bool flipScreen = false;            // Rotate screen 180deg
-String timezone = "";               // custom (non UK) timezone for the clock
-bool hidePlatform = false;          // Hide platform numbers on Rail board?
-int nrTimeOffset = 0;               // Offset minutes for Rail departures display
-int prevUpdateCheckDay;             // Day of the month the last daily firmware update check was made
-unsigned long fwUpdateCheckTimer=0; // Next time to check if the day has rolled over for firmware update check
-bool apiKeys = false;               // Does apikeys.json exist?
-
-char hostname[33];                  // Network hostname (mDNS)
-char myUrl[24];                     // Stores the board's own url
+static unsigned long timer = 0;
+static bool isSleeping = false;            // Is the screen sleeping (showing the "screensaver")
+static bool sleepEnabled = false;          // Is overnight sleep enabled?
+static bool forcedSleep = false;           // Is the system in manual sleep mode?
+static bool forcedAwake = false;           // Was the system woken by touch sensor?
+static int stayAwakeSeconds = 300;         // How long to force stay awake since last tap
+static bool sleepClock = true;             // Showing the clock in sleep mode?
+static bool softResetNeeded = false;       // Is a soft reset pending?
+static bool manualUpdateCheck = false;     // Has the GUI requested a firmware update check
+static bool dateEnabled = false;           // Showing the date on screen?
+static bool weatherEnabled = false;        // Showing weather at station location. Requires an OpenWeatherMap API key.
+static bool enableBus = false;             // Include Bus services on the board?
+static bool firmwareUpdates = true;        // Check for and install firmware updates automatically at boot?
+static bool dailyUpdateCheck = false;      // Check for and install firmware updates at midnight?
+static byte sleepStarts = 0;               // Hour at which the overnight sleep (screensaver) begins
+static byte sleepEnds = 6;                 // Hour at which the overnight sleep (screensaver) ends
+static int brightness = 50;                // Initial brightness level of the OLED screen
+static unsigned long lastWiFiReconnect=0;  // Last WiFi reconnection time (millis)
+static bool firstLoad = true;              // Are we loading for the first time (no station config)?
+static int prevProgressBarPosition=0;      // Used for progress bar smooth animation
+static int startupProgressPercent;         // Initialisation progress
+static bool wifiConnected = false;         // Connected to WiFi?
+static unsigned long nextDataUpdate = 0;   // Next National Rail update time (millis)
+static int dataLoadSuccess = 0;            // Count of successful data downloads
+static int dataLoadFailure = 0;            // Count of failed data downloads
+static unsigned long lastLoadFailure = 0;  // When the last failure occurred
+static int dateWidth;                      // Width of the displayed date in pixels
+static int dateDay;                        // Day of the month of displayed date
+static bool altStationEnabled = false;     // Switch between stations based on time of day
+static bool altStationActive = false;      // Is the alternate station currently shown
+static byte altStarts = 12;                // Hour at which to switch to the alternate station
+static byte altEnds = 23;                  // Hour at which to switch back to the default station
+static bool noScrolling = false;           // Suppress all horizontal scrolling
+static bool flipScreen = false;            // Rotate screen 180deg
+static String timezone = "";               // custom (non UK) timezone for the clock
+static bool hidePlatform = false;          // Hide platform numbers on Rail board?
+static int nrTimeOffset = 0;               // Offset minutes for Rail departures display
+static int prevUpdateCheckDay;             // Day of the month the last daily firmware update check was made
+static unsigned long fwUpdateCheckTimer=0; // Next time to check if the day has rolled over for firmware update check
+static bool apiKeys = false;               // Does apikeys.json exist?
+static bool touchEnabled = false;          // TTP223 Touch Sensor installed?
+static bool touchOverrideAlt = false;      // Touch/Button has switched to/from alternate station
+
+static char hostname[33];                  // Network hostname (mDNS)
+static char myUrl[24];                     // Stores the board's own url
 
 // WiFi Manager status
-bool wifiConfigured = false;        // Is WiFi configured successfully?
+static bool wifiConfigured = false;        // Has WiFi Manager used the captive portal
 
 // Station Board Data
-char nrToken[37] = "";              // National Rail Darwin Lite Tokens are in the format nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn, where each 'n' represents a hexadecimal character (0-9 or a-f).
-char crsCode[4] = "";               // Station code (3 character)
-float stationLat=0;                 // Selected station Latitude/Longitude (used to get weather for the location)
-float stationLon=0;
-char callingCrsCode[4] = "";        // Station code to filter routes on
-char callingStation[45] = "";       // Calling filter station friendly name
-char platformFilter[MAXPLATFORMFILTERSIZE]; // CSV list of platforms to filter on
-char cleanPlatformFilter[MAXPLATFORMFILTERSIZE]; // Cleaned up platform filter (for performance)
-char altCrsCode[4] = "";            // Station code of alternate station
-float altLat=0;                     // Alternate station Latitude/Longitude (used to get weather for the location)
-float altLon=0;
-char altCallingCrsCode[4];          // Station code to filter routes on (when alternate station active)
-char altCallingStation[45] = "";    // Calling filter station friendly name (when alternate station active)
-char altPlatformFilter[MAXPLATFORMFILTERSIZE]; // CSV list of platforms to filter on
-String tflAppkey = "";              // TfL API Key
-char tubeId[13] = "";               // Underground station naptan id
-String tubeName="";                 // Underground Station Name
-char busAtco[13]="";                // Bus Stop ATCO location
-String busName="";                  // Bus Stop long name
-int busDestX;                       // Variable margin for bus destination
-char busFilter[MAXBUSFILTERSIZE]=""; // CSV list of services to filter on
-char cleanBusFilter[MAXBUSFILTERSIZE]; // Cleaned up bus filter (for performance)
-float busLat=0;                     // Bus stop Latitude/Longitude (used to get weather for the location)
-float busLon=0;
+static char nrToken[37] = "";              // National Rail Darwin Lite Tokens are in the format nnnnnnnn-nnnn-nnnn-nnnn-nnnnnnnnnnnn, where each 'n' represents a hexadecimal character (0-9 or a-f).
+static char crsCode[4] = "";               // Station code (3 character)
+static float stationLat=0;                 // Selected station Latitude/Longitude (used to get weather for the location)
+static float stationLon=0;
+static char callingCrsCode[4] = "";        // Station code to filter routes on
+static char callingStation[45] = "";       // Calling filter station friendly name
+static char platformFilter[MAXPLATFORMFILTERSIZE]; // CSV list of platforms to filter on
+static char cleanPlatformFilter[MAXPLATFORMFILTERSIZE]; // Cleaned up platform filter (for performance)
+static char altCrsCode[4] = "";            // Station code of alternate station
+static float altLat=0;                     // Alternate station Latitude/Longitude (used to get weather for the location)
+static float altLon=0;
+static char altCallingCrsCode[4];          // Station code to filter routes on (when alternate station active)
+static char altCallingStation[45] = "";    // Calling filter station friendly name (when alternate station active)
+static char altPlatformFilter[MAXPLATFORMFILTERSIZE]; // CSV list of platforms to filter on
+static String tflAppkey = "";              // TfL API Key
+static char tubeId[13] = "";               // Underground station naptan id
+static String tubeName="";                 // Underground Station Name
+static char busAtco[13]="";                // Bus Stop ATCO location
+static String busName="";                  // Bus Stop long name
+static int busDestX;                       // Variable margin for bus destination
+static char busFilter[MAXBUSFILTERSIZE]=""; // CSV list of services to filter on
+static char cleanBusFilter[MAXBUSFILTERSIZE]; // Cleaned up bus filter (for performance)
+static float busLat=0;                     // Bus stop Latitude/Longitude (used to get weather for the location)
+static float busLon=0;
 
 // board has three possible modes.
 enum boardModes {
+  MODE_LOADCONFIG = -1,
   MODE_RAIL = 0,
   MODE_TUBE = 1,
   MODE_BUS = 2
@@ -384,60 +404,60 @@ static const char standardClassSeating[] PROGMEM = " Standard class seating only
 static const char dualClassSeating[] PROGMEM = " First and Standard class seating available.";
 
 // Animation vars
-int numMessages=0;
-int scrollStopsXpos = 0;
-int scrollStopsYpos = 0;
-int scrollStopsLength = 0;
-bool isScrollingStops = false;
-int currentMessage = 0;
-int prevMessage = 0;
-int prevScrollStopsLength = 0;
-char line2[4+MAXBOARDMESSAGES][MAXCALLINGSIZE+12];
+static int numMessages=0;
+static int scrollStopsXpos = 0;
+static int scrollStopsYpos = 0;
+static int scrollStopsLength = 0;
+static bool isScrollingStops = false;
+static int currentMessage = 0;
+static int prevMessage = 0;
+static int prevScrollStopsLength = 0;
+static char line2[4+MAXBOARDMESSAGES][MAXCALLINGSIZE+12];
 
 // Line 3 (additional services)
-int line3Service = 0;
-int scrollServiceYpos = 0;
-bool isScrollingService = false;
-int prevService = 0;
-bool isShowingVia=false;
-unsigned long serviceTimer=0;
-unsigned long viaTimer=0;
-bool showingMessage = false;
+static int line3Service = 0;
+static int scrollServiceYpos = 0;
+static bool isScrollingService = false;
+static int prevService = 0;
+static bool isShowingVia=false;
+static unsigned long serviceTimer=0;
+static unsigned long viaTimer=0;
+static bool showingMessage = false;
 // TfL/bus specific animation
-int scrollPrimaryYpos = 0;
-bool isScrollingPrimary = false;
-bool attributionScrolled = false;
+static int scrollPrimaryYpos = 0;
+static bool isScrollingPrimary = false;
+static bool attributionScrolled = false;
 
-char displayedTime[29] = "";        // The currently displayed time
-unsigned long nextClockUpdate = 0;  // Next time we need to check/update the clock display
-int fpsDelay=25;                    // Total ms between text movement (for smooth animation)
-unsigned long refreshTimer = 0;
+static char displayedTime[29] = "";        // The currently displayed time
+static unsigned long nextClockUpdate = 0;  // Next time we need to check/update the clock display
+static int fpsDelay=25;                    // Total ms between text movement (for smooth animation)
+static unsigned long refreshTimer = 0;
 
 // Weather Stuff
-char weatherMsg[46];                            // Current weather at station location
-unsigned long nextWeatherUpdate = 0;            // When the next weather update is due
-String openWeatherMapApiKey = "";               // The API key to use
+static char weatherMsg[46];                            // Current weather at station location
+static unsigned long nextWeatherUpdate = 0;            // When the next weather update is due
+static String openWeatherMapApiKey = "";               // The API key to use
 weatherClient currentWeather;                   // Create a weather client
 
 // RSS Client
 rssClient rss;                                  // Create a RSS client
-bool rssEnabled = false;                        // Add RSS feed to the messages
-unsigned long nextRssUpdate = 0;                // When the next RSS update is due
-bool rssAddedtoMsgs = false;
-int lastRssUpdateResult = 0;
-String rssURL;                                  // RSS URL to use
-String rssName;                                 // Name of feed for atrribution
-
-bool noDataLoaded = true;                       // True if no data received for the station
-int lastUpdateResult = 0;                       // Result of last data refresh
-unsigned long lastDataLoadTime = 0;             // Timestamp of last data load
-long apiRefreshRate = DATAUPDATEINTERVAL;       // User selected refresh rate for National Rail API
+static bool rssEnabled = false;                        // Add RSS feed to the messages
+static unsigned long nextRssUpdate = 0;                // When the next RSS update is due
+static bool rssAddedtoMsgs = false;
+static int lastRssUpdateResult = 0;
+static String rssURL;                                  // RSS URL to use
+static String rssName;                                 // Name of feed for atrribution
+
+static bool noDataLoaded = true;                       // True if no data received for the station
+static int lastUpdateResult = 0;                       // Result of last data refresh
+static unsigned long lastDataLoadTime = 0;             // Timestamp of last data load
+static long apiRefreshRate = DATAUPDATEINTERVAL;       // User selected refresh rate for National Rail API
 
 #define MAXHOSTSIZE 48                          // Maximum size of the wsdl Host
 #define MAXAPIURLSIZE 48                        // Maximum size of the wsdl url
 
-char wsdlHost[MAXHOSTSIZE];                     // wsdl Host name
-char wsdlAPI[MAXAPIURLSIZE];                    // wsdl API url
+static char wsdlHost[MAXHOSTSIZE];                     // wsdl Host name
+static char wsdlAPI[MAXAPIURLSIZE];                    // wsdl API url
 
 // RailData XML Client
 raildataXmlClient* raildata = nullptr;
@@ -450,6 +470,9 @@ rdStation station;
 // Station Messages (shared)
 stnMessages messages;
 
+// Optional TTP223 touch sensor / push button
+touchSensor button(34);
+
 /*
  * Graphics helper functions for OLED panel
 */
@@ -534,22 +557,17 @@ void progressBar(const __FlashStringHelper *text, int percent) {
   drawProgressBar(percent);
 }
 
-void drawBuildTime() {
-  char timestamp[22];
-  char buildtime[20];
-  struct tm tm = {};
-
-  sprintf(timestamp,"%s %s",__DATE__,__TIME__);
-  strptime(timestamp,"%b %d %Y %H:%M:%S",&tm);
-  sprintf(buildtime,"v%d.%d-%02d%02d%02d%02d%02d",VERSION_MAJOR,VERSION_MINOR,tm.tm_year-100,tm.tm_mon+1,tm.tm_mday,tm.tm_hour,tm.tm_min);
-  u8g2.drawStr(0,53,buildtime);
+void drawFirmware() {
+  char firmware[16];
+  sprintf(firmware,"B%d.%d-W%d.%d",VERSION_MAJOR,VERSION_MINOR,WEBAPPVER_MAJOR,WEBAPPVER_MINOR);
+   u8g2.drawStr(0,53,firmware);
 }
 
 void drawStartupHeading() {
   u8g2.setFont(NatRailTall12);
   centreText(F("Departures Board"),0);
   u8g2.setFont(NatRailSmall9);
-  drawBuildTime();
+  drawFirmware();
 }
 
 void drawStationHeader(const char *stopName, const char *callingStopName, const char *platFilter, const int timeOffset) {
@@ -891,6 +909,10 @@ void doClockCheck() {
 // Returns true if sleep mode is enabled and we're within the sleep period
 bool isSnoozing() {
   if (forcedSleep) return true;
+  if (forcedAwake) {
+    if (button.secsSinceLastTap() >= stayAwakeSeconds) forcedAwake = false;
+    else return false;
+  }
   if (!sleepEnabled) return false;
   getLocalTime(&timeinfo);
   byte myHour = timeinfo.tm_hour;
@@ -964,12 +986,17 @@ void loadApiKeys() {
   }
 }
 
+void resetLocationIds() {
+  strcpy(crsCode,"");
+  strcpy(tubeId,"");
+  strcpy(busAtco,"");
+}
+
 // Write a default config file so that the Web GUI works initially (force Tube mode if no NR token)
 void writeDefaultConfig() {
     String defaultConfig = "{\"crs\":\"\",\"station\":\"\",\"lat\":0,\"lon\":0,\"weather\":" + String((openWeatherMapApiKey.length())?"true":"false") + F(",\"sleep\":false,\"showDate\":false,\"showBus\":false,\"update\":true,\"sleepStarts\":23,\"sleepEnds\":8,\"brightness\":20,\"tubeId\":\"\",\"tubeName\":\"\",\"mode\":") + String((!nrToken[0])?"1":"0") + "}";
     saveFile(F("/config.json"),defaultConfig);
-    strcpy(crsCode,"");
-    strcpy(tubeId,"");
+    resetLocationIds();
 }
 
 // Load the configuration settings (if they exist, if not create a default set for the Web GUI page to read)
@@ -979,6 +1006,7 @@ void loadConfig() {
   // Set defaults
   strcpy(hostname,defaultHostname);
   timezone = String(ukTimezone);
+  resetLocationIds();
 
   if (LittleFS.exists(F("/config.json"))) {
     File file = LittleFS.open(F("/config.json"), "r");
@@ -1036,6 +1064,8 @@ void loadConfig() {
 
         if (settings[F("noScroll")].is<bool>())          noScrolling = settings[F("noScroll")];
         if (settings[F("flip")].is<bool>())              flipScreen = settings[F("flip")];
+        if (settings[F("touch")].is<bool>())             touchEnabled = settings[F("touch")];
+        if (settings[F("forceWakeTime")].is<int>())      stayAwakeSeconds = settings[F("forceWakeTime")];
         if (settings[F("TZ")].is<const char*>())         timezone = settings[F("TZ")].as<String>();
         if (settings[F("nrTimeOffset")].is<int>())       nrTimeOffset = settings[F("nrTimeOffset")];
         if (settings[F("hidePlatform")].is<bool>())      hidePlatform = settings[F("hidePlatform")];
@@ -1049,12 +1079,12 @@ void loadConfig() {
       }
       file.close();
     }
-  } else if (nrToken[0] || tflAppkey.length()) writeDefaultConfig();
+  } else if (apiKeys) writeDefaultConfig();
 }
 
 // Switch to the alternate station settings if appropriate
-bool setAlternateStation() {
-  if (boardMode==MODE_RAIL && altStationEnabled && isAltActive()) {
+bool setAlternateStation(bool force) {
+  if ((boardMode==MODE_RAIL && altStationEnabled && isAltActive()) || force) {
     // Switch to alternate station
     strcpy(crsCode,altCrsCode);
     stationLat = altLat;
@@ -1074,12 +1104,13 @@ void updateRssFeed() {
 }
 
 // Soft reset/reload
-void softResetBoard() {
-  int previousMode = boardMode;
+void softResetBoard(boardModes forcedMode, bool switchAlt) {
+  boardModes previousMode = boardMode;
   String prevRssUrl = rssURL;
 
   // Reload the settings
   loadConfig();
+  if (forcedMode!=MODE_LOADCONFIG) boardMode = forcedMode;
   if (flipScreen) u8g2.setFlipMode(1); else u8g2.setFlipMode(0);
   if (timezone!="") {
     setenv("TZ",timezone.c_str(),1);
@@ -1141,7 +1172,7 @@ void softResetBoard() {
           int res = raildata->init(wsdlHost, wsdlAPI, &raildataCallback);
           if (res != UPD_SUCCESS) {
             showWsdlFailureScreen();
-             while (true) { server.handleClient(); yield();}
+             while (true) { delay(1);}
           }
         }
         break;
@@ -1170,7 +1201,15 @@ void softResetBoard() {
 
   switch (boardMode) {
     case MODE_RAIL:
-      altStationActive = setAlternateStation();
+      if (switchAlt) {
+        touchOverrideAlt = true;
+        if (!altStationActive) {
+          altStationActive = setAlternateStation(true);
+        }
+      } else {
+        touchOverrideAlt = false;
+        altStationActive = setAlternateStation(false);
+      }
       // Create a cleaned platform filter (if any)
       raildata->cleanFilter(platformFilter,cleanPlatformFilter,sizeof(platformFilter));
       break;
@@ -1189,6 +1228,24 @@ void softResetBoard() {
   messages.numMessages=0;
 }
 
+// Handle switching to next board mode (touch sensor)
+void switchToNextMode() {
+  switch (boardMode) {
+    case MODE_RAIL:
+      if (tubeId[0]) softResetBoard(MODE_TUBE,false);
+      else if (busAtco[0]) softResetBoard(MODE_BUS,false);
+      break;
+    case MODE_TUBE:
+      if (busAtco[0]) softResetBoard(MODE_BUS,false);
+      else if (crsCode[0]) softResetBoard(MODE_RAIL,false);
+      break;
+    case MODE_BUS:
+      if (crsCode[0]) softResetBoard(MODE_RAIL,false);
+      else if (tubeId[0]) softResetBoard(MODE_TUBE,false);
+      break;
+  }
+}
+
 // WiFiManager callback, entered config mode
 void wmConfigModeCallback (WiFiManager *myWiFiManager) {
   showSetupScreen();
@@ -1231,14 +1288,9 @@ bool checkForFirmwareUpdate() {
     return result;
   }
 
-  unsigned long tmr=millis()+1000;
   for (int i=30;i>=0;i--) {
     showFirmwareUpdateWarningScreen(ghUpdate.releaseDescription.c_str(),i);
-    while (tmr>millis()) {
-      yield();
-      server.handleClient();
-    }
-    tmr=millis()+1000;
+    delay(1000);
   }
   u8g2.clearDisplay();
   prevProgressBarPosition=0;
@@ -1331,7 +1383,7 @@ bool getStationBoard() {
     return false;
   } else if (lastUpdateResult == UPD_UNAUTHORISED) {
     showTokenErrorScreen();
-    while (true) { server.handleClient(); yield();}
+    while (true) { delay(1);}
   } else {
     showUpdateIcon(false);
     dataLoadFailure++;
@@ -1622,7 +1674,7 @@ bool getUndergroundBoard() {
     return false;
   } else if (lastUpdateResult == UPD_UNAUTHORISED) {
     showTokenErrorScreen();
-    while (true) { server.handleClient(); yield();}
+    while (true) { delay(1);}
   } else {
     showUpdateIcon(false);
     dataLoadFailure++;
@@ -1728,7 +1780,7 @@ bool getBusDeparturesBoard() {
     return false;
   } else if (lastUpdateResult == UPD_UNAUTHORISED) {
     showTokenErrorScreen();
-    while (true) { server.handleClient(); yield();}
+    while (true) { delay(1);}
   } else {
     showUpdateIcon(false);
     dataLoadFailure++;
@@ -1818,19 +1870,17 @@ void drawBusDeparturesBoard() {
  */
 
 // Helper function for returning text status messages
-void sendResponse(int code, const __FlashStringHelper* msg) {
-  server.send(code, contentTypeText, msg);
+void sendResponse(int code, const __FlashStringHelper* msg, AsyncWebServerRequest *request) {
+  request->send(code,contentTypeText,msg);
 }
 
-void sendResponse(int code, String msg) {
-  server.send(code, contentTypeText, msg);
+void sendResponse(int code, String msg, AsyncWebServerRequest *request) {
+  request->send(code,contentTypeText,msg);
 }
 
 // Return the correct MIME type for a file name
 String getContentType(String filename) {
-  if (server.hasArg(F("download"))) {
-    return F("application/octet-stream");
-  } else if (filename.endsWith(F(".htm"))) {
+  if (filename.endsWith(F(".htm"))) {
     return F("text/html");
   } else if (filename.endsWith(F(".html"))) {
     return F("text/html");
@@ -1865,135 +1915,25 @@ String getContentType(String filename) {
 }
 
 // Stream a file from the file system
-bool handleStreamFile(String filename) {
+bool handleStreamFile(String filename, AsyncWebServerRequest *request) {
   if (LittleFS.exists(filename)) {
-    File file = LittleFS.open(filename,"r");
     String contentType = getContentType(filename);
-    server.streamFile(file, contentType);
-    file.close();
+    request->send(LittleFS,filename,contentType);
     return true;
   } else return false;
 }
 
 // Stream a file stored in PROGMEM flash (default graphics are now included in the firmware image)
-void handleStreamFlashFile(String filename, const uint8_t *filedata, size_t contentLength) {
-
+void handleStreamFlashFile(String filename, const uint8_t *filedata, size_t contentLength, AsyncWebServerRequest *request) {
   String contentType = getContentType(filename);
-  WiFiClient client = server.client();
-  // Send the headers
-  client.println(F("HTTP/1.1 200 OK"));
-  client.print(F("Content-Type: "));
-  client.println(contentType);
-  client.print(F("Content-Length: "));
-  client.println(contentLength);
-  client.println(F("Connection: close"));
-  client.println(); // End of headers
-
-  const size_t chunkSize = 512;
-  uint8_t buffer[chunkSize];
-  size_t sent = 0;
-
-  while (sent < contentLength) {
-    size_t toSend = min(chunkSize, contentLength - sent);
-    // Copy from PROGMEM to buffer
-    for (size_t i=0;i<toSend;i++) {
-      buffer[i] = pgm_read_byte(&filedata[sent + i]);
-    }
-    client.write(buffer, toSend);
-    sent += toSend;
-  }
-}
-
-// Save the API keys POSTed from the keys.htm page
-// If an OWM key is passed, this is tested before being committed to the file system. It's not possible
-// to check the National Rail or TfL tokens at this point.
-//
-void handleSaveKeys() {
-  String newJSON, owmToken, nrToken;
-  JsonDocument doc;
-  bool result = true;
-  String msg = F("API keys saved successfully.");
-
-  if ((server.method() == HTTP_POST) && (server.hasArg("plain"))) {
-    newJSON = server.arg("plain");
-    // Deserialise to get the OWM API key
-    DeserializationError error = deserializeJson(doc, newJSON);
-    if (!error) {
-      JsonObject settings = doc.as<JsonObject>();
-      if (settings[F("owmToken")].is<const char*>()) {
-        owmToken = settings[F("owmToken")].as<String>();
-        if (owmToken.length()) {
-          // Check if this is a valid token...
-          if (!currentWeather.updateWeather(owmToken, "51.52", "-0.13")) {
-            msg = F("The OpenWeather Map API key is not valid. Please check you have copied your key correctly. It may take up to 30 minutes for a newly created key to become active.\n\nNo changes have been saved.");
-            result = false;
-          }
-        }
-      }
-      if (result) {
-        if (!saveFile(F("/apikeys.json"),newJSON)) {
-          msg = F("Failed to save the API keys to the file system (file system corrupt or full?)");
-          result = false;
-        } else {
-          nrToken = settings[F("nrToken")].as<String>();
-          if (!nrToken.length()) msg+=F("\n\nNote: Only Tube and Bus Departures will be available without a National Rail token.");
-        }
-      }
-    } else {
-      msg = F("Invalid JSON format. No changes have been saved.");
-      result = false;
-    }
-    if (result) {
-      // Load/Update the API Keys in memory
-      loadApiKeys();
-      // If all location codes are blank we're in the setup process. If not, the keys have been changed so just reboot.
-      if (!crsCode[0] && !tubeId[0] && !busAtco[0]) {
-        sendResponse(200,msg);
-        writeDefaultConfig();
-        showSetupCrsHelpScreen();
-      } else {
-        msg += F("\n\nThe system will now restart.");
-        sendResponse(200,msg);
-        delay(500);
-        ESP.restart();
-      }
-    } else {
-      sendResponse(400,msg);
-    }
-  } else {
-    sendResponse(400,F("Invalid"));
-  }
-}
-
-// Save configuration setting POSTed from index.htm
-void handleSaveSettings() {
-  String newJSON;
-
-  if ((server.method() == HTTP_POST) && (server.hasArg("plain"))) {
-    newJSON = server.arg("plain");
-    saveFile(F("/config.json"),newJSON);
-    if ((!crsCode[0] && !tubeId[0]) || server.hasArg("reboot")) {
-      // First time setup or base config change, we need a full reboot
-      sendResponse(200,F("Configuration saved. The system will now restart."));
-      delay(1000);
-      ESP.restart();
-    } else {
-      sendResponse(200,F("Configuration updated. The system will update shortly."));
-      softResetBoard();
-    }
-  } else {
-    // Something went wrong saving the config file
-    sendResponse(400,F("The configuration could not be updated. The system will restart."));
-    delay(1000);
-    ESP.restart();
-  }
+  request->send(200,contentType,(uint8_t *)filedata, contentLength);
 }
 
 /*
  * Expose the file system via the Web GUI with some basic functions for directory browsing, file reading and deletion.
  */
 
- // Return storage information
+// Return storage information
 String getFSInfo() {
   char info[70];
 
@@ -2002,9 +1942,9 @@ String getFSInfo() {
 }
 
 // Send a basic directory listing to the browser
-void handleFileList() {
+void handleFileList(AsyncWebServerRequest *request) {
   String path;
-  if (!server.hasArg("dir")) path="/"; else path = server.arg("dir");
+  if (!request->hasParam("dir")) path="/"; else path = request->getParam("dir")->value();
   File root = LittleFS.open(path);
 
   String output=F("<html><body style=\"font-family:Helvetica,Arial,sans-serif\"><h2>Departures Board File System</h2>");
@@ -2028,70 +1968,37 @@ void handleFileList() {
 
   output += F("</table><br>");
   output += getFSInfo() + F("<p><a href=\"/upload\">Upload</a> a file</p></body></html>");
-  server.send(200,contentTypeHtml,output);
+  request->send(200,contentTypeHtml,output);
 }
 
 // Stream a file to the browser
-void handleCat() {
-  String filename;
-
-  if (server.hasArg(F("f"))) {
-    handleStreamFile(server.arg("f"));
-  } else sendResponse(404,F("Not found"));
+void handleCat(AsyncWebServerRequest *request) {
+  if (request->hasParam("f")) {
+    String filename = request->getParam("f")->value();
+    handleStreamFile(filename,request);
+  } else sendResponse(404,F("Not found"),request);
 }
 
 // Delete a file from the file system
-void handleDelete() {
-  String filename;
-
-  if (server.hasArg(F("f"))) {
-    if (LittleFS.remove(server.arg(F("f")))) {
+void handleDelete(AsyncWebServerRequest *request) {
+  if (request->hasParam("f")) {
+    String filename = request->getParam("f")->value();
+    if (LittleFS.remove(filename)) {
       // Successfully removed go back to directory listing
-      server.sendHeader(F("Location"),F("/dir"));
-      server.send(303);
-    } else sendResponse(400,F("Failed to delete file"));
-  } else sendResponse(404,F("Not found"));
-
+      request->redirect("/dir");
+    } else sendResponse(400,F("Failed to delete file"),request);
+  } else sendResponse(404,F("Not found"),request);
 }
 
 // Format the file system
-void handleFormatFFS() {
+void handleFormatFFS(AsyncWebServerRequest *request) {
   String message;
 
   if (LittleFS.format()) {
     message=F("File System was successfully formatted\n\n");
     message+=getFSInfo();
   } else message=F("File System could not be formatted!");
-  sendResponse(200,message);
-}
-
-// Upload a file from the browser
-void handleFileUpload() {
-  HTTPUpload& upload = server.upload();
-  if (upload.status == UPLOAD_FILE_START) {
-    String filename = upload.filename;
-    if (!filename.startsWith(F("/"))) {
-      filename = "/" + filename;
-    }
-    fsUploadFile = LittleFS.open(filename, "w");
-    filename = String();
-  } else if (upload.status == UPLOAD_FILE_WRITE) {
-    if (fsUploadFile) {
-      fsUploadFile.write(upload.buf, upload.currentSize);
-    }
-  } else if (upload.status == UPLOAD_FILE_END) {
-    WiFiClient client = server.client();
-    if (fsUploadFile) {
-      fsUploadFile.close();
-      client.println(F("HTTP/1.1 302 Found"));
-      client.println(F("Location: /success"));
-      client.println(F("Connection: close"));
-    } else {
-      client.println(F("HTTP/1.1 500 Could not create file"));
-      client.println(F("Connection: close"));
-    }
-    client.println();
-  }
+  sendResponse(200,message,request);
 }
 
 /*
@@ -2099,18 +2006,19 @@ void handleFileUpload() {
  */
 
 // Fallback function for browser requests
-void handleNotFound() {
-  if ((LittleFS.exists(server.uri())) && (server.method() == HTTP_GET)) handleStreamFile(server.uri());
-  else if (server.uri() == F("/keys.htm")) handleStreamFlashFile(server.uri(), keyshtm, sizeof(keyshtm));
-  else if (server.uri() == F("/index.htm")) handleStreamFlashFile(server.uri(), indexhtm, sizeof(indexhtm));
-  else if (server.uri() == F("/nrelogo.webp")) handleStreamFlashFile(server.uri(), nrelogo, sizeof(nrelogo));
-  else if (server.uri() == F("/tfllogo.webp")) handleStreamFlashFile(server.uri(), tfllogo, sizeof(tfllogo));
-  else if (server.uri() == F("/btlogo.webp")) handleStreamFlashFile(server.uri(), btlogo, sizeof(btlogo));
-  else if (server.uri() == F("/tube.webp")) handleStreamFlashFile(server.uri(), tubeicon, sizeof(tubeicon));
-  else if (server.uri() == F("/nr.webp")) handleStreamFlashFile(server.uri(), nricon, sizeof(nricon));
-  else if (server.uri() == F("/favicon.png")) handleStreamFlashFile(server.uri(), faviconpng, sizeof(faviconpng));
-  else if (server.uri() == F("/rss.json")) handleStreamFlashFile(server.uri(), rssjson, sizeof(rssjson));
-  else sendResponse(404,F("Not Found"));
+void handleNotFound(AsyncWebServerRequest *request) {
+  if ((LittleFS.exists(request->url())) && (request->method() == HTTP_GET)) handleStreamFile(request->url(),request);
+  else if (request->url() == F("/keys.htm")) handleStreamFlashFile(request->url(), keyshtm, sizeof(keyshtm),request);
+  else if (request->url() == F("/index.htm")) handleStreamFlashFile(request->url(), indexhtm, sizeof(indexhtm),request);
+  else if (request->url() == F("/editrss.htm")) handleStreamFlashFile(request->url(), editrsshtm, sizeof(editrsshtm),request);
+  else if (request->url() == F("/nrelogo.webp")) handleStreamFlashFile(request->url(), nrelogo, sizeof(nrelogo),request);
+  else if (request->url() == F("/tfllogo.webp")) handleStreamFlashFile(request->url(), tfllogo, sizeof(tfllogo),request);
+  else if (request->url() == F("/btlogo.webp")) handleStreamFlashFile(request->url(), btlogo, sizeof(btlogo),request);
+  else if (request->url() == F("/tube.webp")) handleStreamFlashFile(request->url(), tubeicon, sizeof(tubeicon),request);
+  else if (request->url() == F("/nr.webp")) handleStreamFlashFile(request->url(), nricon, sizeof(nricon),request);
+  else if (request->url() == F("/favicon.png")) handleStreamFlashFile(request->url(), faviconpng, sizeof(faviconpng),request);
+  else if (request->url() == F("/rss.json")) handleStreamFlashFile(request->url(), rssjson, sizeof(rssjson),request);
+  else sendResponse(404,F("Not Found"),request);
 }
 
 String getResultCodeText(int resultCode) {
@@ -2146,7 +2054,7 @@ String getResultCodeText(int resultCode) {
 }
 
 // Send some useful system & station information to the browser
-void handleInfo() {
+void handleInfo(AsyncWebServerRequest *request) {
   unsigned long uptime = millis();
   char sysUptime[30];
   int days = uptime / msDay ;
@@ -2193,71 +2101,63 @@ void handleInfo() {
   if (weatherEnabled) {
     message+="Last weather result: " + currentWeather.lastErrorMsg + F("\nNext weather update: ") + String(nextWeatherUpdate-millis()) + F("ms");
   }
-  sendResponse(200,message);
+  sendResponse(200,message,request);
 }
 
 // Stream the index.htm page unless we're in first time setup and need the api keys
-void handleRoot() {
+void handleRoot(AsyncWebServerRequest *request) {
   if (!apiKeys) {
-    if (LittleFS.exists(F("/keys.htm"))) handleStreamFile(F("/keys.htm")); else handleStreamFlashFile(F("/keys.htm"),keyshtm,sizeof(keyshtm));
+    if (LittleFS.exists(F("/keys.htm"))) handleStreamFile(F("/keys.htm"),request); else handleStreamFlashFile(F("/keys.htm"),keyshtm,sizeof(keyshtm),request);
   } else {
-    if (LittleFS.exists(F("/index_d.htm"))) handleStreamFile(F("/index_d.htm")); else handleStreamFlashFile(F("/index.htm"),indexhtm,sizeof(indexhtm));
+    if (LittleFS.exists(F("/index_d.htm"))) handleStreamFile(F("/index_d.htm"),request); else handleStreamFlashFile(F("/index.htm"),indexhtm,sizeof(indexhtm),request);
   }
 }
 
 // Send the firmware version to the client (called from index.htm)
-void handleFirmwareInfo() {
-  String response = "{\"firmware\":\"B" + String(VERSION_MAJOR) + "." + String(VERSION_MINOR) + "-W" + String(WEBAPPVER_MAJOR) + "." + String(WEBAPPVER_MINOR) + F(" Build:") + getBuildTime() + F("\"}");
-  server.send(200,contentTypeJson,response);
+void handleFirmwareInfo(AsyncWebServerRequest *request) {
+  String response = "{\"firmware\":\"B" + String(VERSION_MAJOR) + "." + String(VERSION_MINOR) + "-W" + String(WEBAPPVER_MAJOR) + "." + String(WEBAPPVER_MINOR) + F("\"}");
+  request->send(200,contentTypeJson,response);
 }
 
 // Force a reboot of the ESP32
-void handleReboot() {
-  sendResponse(200,F("The Departures Board is restarting..."));
-  delay(1000);
-  ESP.restart();
+void handleReboot(AsyncWebServerRequest *request) {
+  sendResponse(200,F("The Departures Board is restarting..."),request);
+  restartTimer.once(1, []() { ESP.restart(); });
 }
 
 // Erase the stored WiFiManager credentials
-void handleEraseWiFi() {
-  sendResponse(200,F("Erasing stored WiFi. You will need to connect to the \"Departures Board\" network and use WiFi Manager to reconfigure."));
-  delay(1000);
-  WiFiManager wm;
-  wm.resetSettings();
-  delay(500);
-  ESP.restart();
+void handleEraseWiFi(AsyncWebServerRequest *request) {
+  sendResponse(200,F("Erasing stored WiFi. You will need to connect to the \"Departures Board\" network and use WiFi Manager to reconfigure."),request);
+  restartTimer.once(1, []() { WiFiManager wm; wm.resetSettings(); ESP.restart();});
 }
 
 // "Factory reset" the app - delete WiFi, format file system and reboot
-void handleFactoryReset() {
-  sendResponse(200,F("Factory reseting the Departures Board..."));
-  delay(1000);
-  WiFiManager wm;
-  wm.resetSettings();
-  delay(500);
-  LittleFS.format();
-  delay(500);
-  ESP.restart();
+void handleFactoryReset(AsyncWebServerRequest *request) {
+  sendResponse(200,F("Factory reseting the Departures Board..."),request);
+  restartTimer.once(1, []() { WiFiManager wm; wm.resetSettings(); LittleFS.format(); ESP.restart();});
 }
 
 // Interactively change the brightness of the OLED panel (called from index.htm)
-void handleBrightness() {
-  if (server.hasArg(F("b"))) {
-    int level = server.arg(F("b")).toInt();
+void handleBrightness(AsyncWebServerRequest *request) {
+  if (request->hasParam("b")) {
+    int level = request->getParam("b")->value().toInt();
     if (level>0 && level<256) {
       u8g2.setContrast(level);
       brightness = level;
-      sendResponse(200,F("OK"));
+      sendResponse(200,F("OK"),request);
       return;
     }
   }
-  sendResponse(200,F("invalid request"));
+  sendResponse(200,F("invalid request"),request);
 }
 
 // Web GUI has requested updates be installed
-void handleOtaUpdate() {
-  sendResponse(200,F("Update initiated - check Departure Board display for progress"));
-  delay(500);
+void handleOtaUpdate(AsyncWebServerRequest *request) {
+  sendResponse(200,F("Update initiated - check Departure Board display for progress"),request);
+  manualUpdateCheck = true;
+}
+
+void doManualOtaCheck() {
   u8g2.clearBuffer();
   u8g2.setFont(NatRailTall12);
   centreText(F("Getting latest firmware details from GitHub..."),26);
@@ -2270,26 +2170,25 @@ void handleOtaUpdate() {
       showUpdateCompleteScreen("Firmware Update Check Failed","Unable to retrieve latest release information.",ghUpdate.getLastError().c_str(),"",i,false);
       delay(1000);
     }
-    log_e("FW Update failed: %s\n",ghUpdate.getLastError().c_str());
   }
   // Always restart
   ESP.restart();
 }
 
 // Endpoint for controlling sleep mode
-void handleControl() {
+void handleControl(AsyncWebServerRequest *request) {
   String resp = "{\"sleeping\":";
-  if (server.hasArg(F("sleep"))) {
-    if (server.arg(F("sleep")) == "1") forcedSleep=true; else forcedSleep=false;
+  if (request->hasParam("sleep")) {
+    if (request->getParam("sleep")->value() == "1") forcedSleep=true; else forcedSleep=false;
   }
-  if (server.hasArg(F("clock"))) {
-    if (server.arg(F("clock")) == "1") sleepClock=true; else sleepClock=false;
+  if (request->hasParam("clock")) {
+    if (request->getParam("clock")->value() == "1") sleepClock=true; else sleepClock=false;
   }
   resp += (isSleeping || forcedSleep) ? "true":"false";
   resp += F(",\"display\":");
   resp += (sleepClock || (!isSleeping && !forcedSleep)) ? "true":"false";
   resp += "}";
-  server.send(200, contentTypeJson, resp);
+  request->send(200, contentTypeJson, resp);
 }
 
 /*
@@ -2297,99 +2196,74 @@ void handleControl() {
  */
 
 // Call the National Rail Station Picker (called from index.htm)
-void handleStationPicker() {
-  if (!server.hasArg(F("q"))) {
-    sendResponse(400, F("Missing Query"));
+void handleStationPicker(AsyncWebServerRequest *request)
+{
+  if (!request->hasParam("q")) {
+    sendResponse(400,F("Missing Query"),request);
     return;
   }
 
-  String query = server.arg(F("q"));
+  String query = request->getParam("q")->value();
   if (query.length() <= 2) {
-    sendResponse(400, F("Query Too Short"));
+    sendResponse(400,F("Query too short"),request);
     return;
   }
 
   const char* host = "stationpicker.nationalrail.co.uk";
-  WiFiClientSecure httpsClient;
-  httpsClient.setInsecure();
-  httpsClient.setTimeout(10000);
-
-  int retryCounter = 0;
-  while (!httpsClient.connect(host, 443) && retryCounter++ < 20) {
-    delay(50);
-  }
+  WiFiClientSecure client;
+  client.setInsecure();
+  client.setTimeout(4000);
 
-  if (retryCounter >= 20) {
-    sendResponse(408, F("NR Timeout"));
+  if (!client.connect(host, 443)) {
+    sendResponse(408, F("NR Connect Timeout"),request);
     return;
   }
 
-  httpsClient.print(String("GET /stationPicker/") + query + F(" HTTP/1.0\r\n") +
-                    F("Host: stationpicker.nationalrail.co.uk\r\n") +
-                    F("Referer: https://www.nationalrail.co.uk\r\n") +
-                    F("Origin: https://www.nationalrail.co.uk\r\n") +
-                    F("Connection: close\r\n\r\n"));
+  client.print(String("GET /stationPicker/") + query + " HTTP/1.0\r\n"
+               "Host: stationpicker.nationalrail.co.uk\r\n"
+               "Referer: https://www.nationalrail.co.uk\r\n"
+               "Origin: https://www.nationalrail.co.uk\r\n"
+               "Connection: close\r\n\r\n");
 
-  // Wait for response header
-  retryCounter = 0;
-  while (!httpsClient.available() && retryCounter++ < 15) {
-    delay(100);
+  int requestTimer = 0;
+  while (!client.available() && requestTimer<1000) {
+    requestTimer++;
+    delay(1);
   }
 
-  if (!httpsClient.available()) {
-    httpsClient.stop();
-    sendResponse(408, F("NRQ Timeout"));
-    return;
+  if (!client.available()) {
+    client.stop();
+    sendResponse(408,F("NRQ Timeout"),request);
   }
 
-  // Parse status code
-  String statusLine = httpsClient.readStringUntil('\n');
-  if (!statusLine.startsWith(F("HTTP/")) || statusLine.indexOf(F("200 OK")) == -1) {
-    httpsClient.stop();
+  String statusLine = client.readStringUntil('\n');
 
-    if (statusLine.indexOf(F("401")) > 0) {
-      sendResponse(401, F("Not Authorized"));
-    } else if (statusLine.indexOf(F("500")) > 0) {
-      sendResponse(500, F("Server Error"));
-    } else {
-      sendResponse(503, statusLine.c_str());
-    }
+  if (statusLine.indexOf("200") == -1) {
+    client.stop();
+    sendResponse(503, statusLine, request);
     return;
   }
 
   // Skip the remaining headers
-  while (httpsClient.connected() || httpsClient.available()) {
-    String line = httpsClient.readStringUntil('\n');
+  while (client.connected() || client.available()) {
+    String line = client.readStringUntil('\n');
     if (line == "\r") break;
   }
 
   // Start sending response
-  server.setContentLength(CONTENT_LENGTH_UNKNOWN);
-  server.send(200, contentTypeJson, "");
-
-  String buffer;
+  AsyncResponseStream *response = request->beginResponseStream("application/json");
+  uint8_t buffer[512];
   unsigned long timeout = millis() + 5000UL;
-
-  while ((httpsClient.connected() || httpsClient.available()) && millis() < timeout) {
-    while (httpsClient.available()) {
-      char c = httpsClient.read();
-      if (c <= 128) buffer += c;
-      if (buffer.length() >= 1024) {
-        server.sendContent(buffer);
-        buffer = "";
-        yield();
-      }
+  while ((client.connected() || client.available()) && millis() < timeout) {
+    int len = client.read(buffer, sizeof(buffer));
+    if (len > 0) {
+      response->write(buffer, len);
+      delay(1);
     }
   }
 
-  // Flush remaining buffer
-  if (buffer.length()) {
-    server.sendContent(buffer);
-  }
-
-  httpsClient.stop();
-  server.sendContent("");
-  server.client().stop();
+  client.stop();
+  request->send(response);
 }
 
 // Update the current weather message if weather updates are enabled and we have a lat/lon for the selected location
@@ -2416,7 +2290,7 @@ void updateCurrentWeather(float latitude, float longitude) {
 //
 void departureBoardLoop() {
 
-  if (altStationEnabled && !isSleeping && altStationActive != isAltActive()) softResetBoard(); // Switch between station views
+  if (altStationEnabled && !isSleeping && !touchOverrideAlt && altStationActive != isAltActive()) softResetBoard(MODE_LOADCONFIG,false); // Switch between station views
 
   if ((millis() > nextDataUpdate) && (!isScrollingStops) && (!isScrollingService) && (lastUpdateResult != UPD_UNAUTHORISED) && (!isSleeping) && (wifiConnected)) {
     timer = millis() + 2000;
@@ -2829,16 +2703,17 @@ void setup(void) {
   WiFi.setSleep(WIFI_PS_NONE);      // Turn off WiFi Powersaving
   WiFi.hostname(hostname);          // Set the hostname ("Departures Board")
   WiFi.mode(WIFI_STA);              // Enter WiFi station mode
-  WiFiManager wm;                   // Start WiFiManager
+
+  WiFiManager wm;                             // Start WiFiManager
   wm.setAPCallback(wmConfigModeCallback);     // Set the callback for config mode notification
   wm.setWiFiAutoReconnect(true);              // Attempt to auto-reconnect WiFi
   wm.setConnectTimeout(8);
   wm.setConnectRetries(2);
 
   bool result = wm.autoConnect("Departures Board");    // Attempt to connect to WiFi (or enter interactive configuration mode)
-  if (!result) {
-      // Failed to connect/configure
-      ESP.restart();
+  if (!result || wifiConfigured) {
+    // Need to restart after config (cannot reuse port)
+    ESP.restart();
   }
 
   // Wait for WiFi connection
@@ -2863,60 +2738,197 @@ void setup(void) {
   u8g2.sendBuffer();                                              // Send to OLED panel
 
   // Configure the local webserver paths
-  server.on(F("/"),handleRoot);
-  server.on(F("/erasewifi"),handleEraseWiFi);
-  server.on(F("/factoryreset"),handleFactoryReset);
-  server.on(F("/info"),handleInfo);
-  server.on(F("/formatffs"),handleFormatFFS);
-  server.on(F("/dir"),handleFileList);
-  server.onNotFound(handleNotFound);
-  server.on(F("/cat"),handleCat);
-  server.on(F("/del"),handleDelete);
-  server.on(F("/reboot"),handleReboot);
-  server.on(F("/stationpicker"),handleStationPicker);           // Used by the Web GUI to lookup station codes interactively
-  server.on(F("/firmware"),handleFirmwareInfo);                 // Used by the Web GUI to display the running firmware version
-  server.on(F("/savesettings"),HTTP_POST,handleSaveSettings);   // Used by the Web GUI to save updated configuration settings
-  server.on(F("/savekeys"),HTTP_POST,handleSaveKeys);           // Used by the Web GUI to verify/save API keys
-  server.on(F("/brightness"),handleBrightness);                 // Used by the Web GUI to interactively set the panel brightness
-  server.on(F("/ota"),handleOtaUpdate);                         // Used by the Web GUI to initiate a manual firmware/WebApp update
-  server.on(F("/control"),handleControl);                       // Endpoint for automation
-
-  server.on(F("/update"), HTTP_GET, []() {
-    server.sendHeader("Connection", "close");
-    server.send(200, contentTypeHtml, updatePage);
-  });
-  /*handling uploading firmware file */
-  server.on(F("/update"), HTTP_POST, []() {
-    server.sendHeader("Connection", "close");
-    sendResponse(200,(Update.hasError()) ? "FAIL" : "OK");
-    ESP.restart();
-  }, []() {
-    HTTPUpload& upload = server.upload();
-    if (upload.status == UPLOAD_FILE_START) {
-      if (!Update.begin(UPDATE_SIZE_UNKNOWN)) { //start with max available size
-        //Update.printError(Serial);
+  server.on("/", HTTP_GET, [](AsyncWebServerRequest *request){handleRoot(request);});
+  server.on("/erasewifi", HTTP_GET, [](AsyncWebServerRequest *request){handleEraseWiFi(request);});
+  server.on("/factoryreset", HTTP_GET, [](AsyncWebServerRequest *request){handleFactoryReset(request);});
+  server.on("/info", HTTP_GET, [](AsyncWebServerRequest *request){handleInfo(request);});
+  server.on("/formatffs", HTTP_GET, [](AsyncWebServerRequest *request){handleFormatFFS(request);});
+  server.on("/dir", HTTP_GET, [](AsyncWebServerRequest *request){handleFileList(request);});
+  server.onNotFound([](AsyncWebServerRequest *request){handleNotFound(request);});
+  server.on("/cat", HTTP_GET, [](AsyncWebServerRequest *request){handleCat(request);});
+  server.on("/del", HTTP_GET, [](AsyncWebServerRequest *request){handleDelete(request);});
+  server.on("/reboot", HTTP_GET, [](AsyncWebServerRequest *request){handleReboot(request);});
+  server.on("/stationpicker", HTTP_GET, [](AsyncWebServerRequest *request){handleStationPicker(request);});
+  server.on("/firmware", HTTP_GET, [](AsyncWebServerRequest *request){handleFirmwareInfo(request);});
+  server.on("/brightness", HTTP_GET, [](AsyncWebServerRequest *request){handleBrightness(request);});
+  server.on("/ota", HTTP_GET, [](AsyncWebServerRequest *request){handleOtaUpdate(request);});
+  server.on("/control", HTTP_GET, [](AsyncWebServerRequest *request){handleControl(request);});
+  server.on("/success", HTTP_GET, [](AsyncWebServerRequest *request){request->send(200,contentTypeHtml,successPage);});
+
+  //
+  // Save settings returned by the Web GUI
+  //
+  server.on("/savesettings", HTTP_POST, [](AsyncWebServerRequest *request) {
+    if (request->_tempObject) {
+      String* body = (String*)(request->_tempObject);
+      saveFile(F("/config.json"), body->c_str());
+
+      delete body; // Clean up memory
+      request->_tempObject = nullptr;
+
+      if ((!crsCode[0] && !tubeId[0]) || request->hasParam("reboot")) {
+        // First time setup or base config change, we need a full reboot
+        sendResponse(200,F("Configuration saved. The system will now restart."),request);
+        restartTimer.once(1, []() { ESP.restart(); });
+      } else {
+        sendResponse(200,F("Configuration updated. The system will update shortly."),request);
+        softResetNeeded = true;
       }
-    } else if (upload.status == UPLOAD_FILE_WRITE) {
-      /* flashing firmware to ESP*/
-      if (Update.write(upload.buf, upload.currentSize) != upload.currentSize) {
-        //Update.printError(Serial);
+    } else {
+      sendResponse(400,F("Empty"),request);
+    }
+  }, NULL, [](AsyncWebServerRequest *request, uint8_t *data, size_t len, size_t index, size_t total) {
+    if (!index) {
+      // First chunk: Create a String object in RAM
+      request->_tempObject = new String("");
+    }
+
+    String* body = (String*)(request->_tempObject);
+    for (size_t i = 0; i < len; i++) {
+      body->concat((char)data[i]);
+    }
+  });
+
+  //
+  // Save the API keys returned from the Web GUI
+  //
+  server.on("/savekeys", HTTP_POST, [](AsyncWebServerRequest *request) {
+    if (request->_tempObject) {
+      String* body = (String*)(request->_tempObject);
+
+      JsonDocument doc;
+      bool result = true;
+      String msg = F("API keys saved successfully.");
+      DeserializationError error = deserializeJson(doc, body->c_str());
+      if (!error) {
+        if (!saveFile(F("/apikeys.json"), body->c_str())) {
+          msg = F("Failed to save the API keys to the file system (file system corrupt or full?)");
+          result = false;
+        } else {
+          JsonObject settings = doc.as<JsonObject>();
+          String nrToken = settings[F("nrToken")].as<String>();
+          if (!nrToken.length()) msg+=F("\n\nNote: Only Tube and Bus Departures will be available without a National Rail token.");
+        }
+      } else {
+        msg = F("Invalid JSON format. No changes have been saved.");
+        result = false;
       }
-    } else if (upload.status == UPLOAD_FILE_END) {
-      if (Update.end(true)) { //true to set the size to the current progress
+
+      delete body; // Clean up memory
+      request->_tempObject = nullptr;
+
+      if (result) {
+        // Load/Update the API Keys in memory
+        loadApiKeys();
+        // If all location codes are blank we're in the setup process. If not, the keys have been changed so just reboot.
+        if (!crsCode[0] && !tubeId[0] && !busAtco[0]) {
+          sendResponse(200,msg,request);
+          writeDefaultConfig();
+          showSetupCrsHelpScreen();
+        } else {
+          msg += F("\n\nThe system will now restart.");
+          sendResponse(200,msg,request);
+          restartTimer.once(1, []() { ESP.restart(); });
+        }
       } else {
-        //Update.printError(Serial);
+        sendResponse(400,msg,request);
       }
+    } else {
+      sendResponse(400,F("Empty"),request);
+    }
+  }, NULL, [](AsyncWebServerRequest *request, uint8_t *data, size_t len, size_t index, size_t total) {
+    if (!index) {
+      // First chunk: Create a String object in RAM
+      request->_tempObject = new String("");
+    }
+
+    String* body = (String*)(request->_tempObject);
+    for (size_t i = 0; i < len; i++) {
+      body->concat((char)data[i]);
     }
   });
 
-  server.on(F("/upload"), HTTP_GET, []() {
-      server.send(200, contentTypeHtml, uploadPage);
+  //
+  // Handle uploads to LittleFS
+  //
+  server.on("/upload", HTTP_GET, [](AsyncWebServerRequest *request){request->send(200,contentTypeHtml,uploadPage);});
+  server.on("/upload", HTTP_POST, [](AsyncWebServerRequest *request) {
+    request->redirect("/success");
+  }, [](AsyncWebServerRequest *request, String filename, size_t index, uint8_t *data, size_t len, bool final) {
+    if (!index) {
+      String path = "/" + filename;
+      if (LittleFS.exists(path)) LittleFS.remove(path);
+      size_t fileSize = request->header("Content-Length").toInt();
+      size_t availableSpace = LittleFS.totalBytes() - LittleFS.usedBytes() - 1024;
+
+      if (fileSize > availableSpace) {
+          sendResponse(507,F("Insufficient storage space in File System"),request);
+          request->client()->close();
+          return;
+      }
+      // First chunk: Create/Open the file and store the handle in _tempObject
+      // We use a pointer to a File object so we can keep it open between chunks
+      File *file = new File(LittleFS.open(path, FILE_WRITE));
+      if (!*file) {
+        sendResponse(500,F("File System Error"),request);
+        request->client()->close();
+        return;
+      }
+      request->_tempObject = file;
+    }
+
+    // If we have a valid file handle, write the current chunk
+    if (len && request->_tempObject) {
+      File *file = reinterpret_cast<File *>(request->_tempObject);
+      file->write(data, len);
+    }
+
+    if (final && request->_tempObject) {
+      // Last chunk: Close the file and clean up the pointer
+      File *file = reinterpret_cast<File *>(request->_tempObject);
+      file->close();
+      delete file;
+      request->_tempObject = nullptr;
+    }
   });
-  server.on(F("/upload"), HTTP_POST, []() {
-  }, handleFileUpload);
 
-  server.on(F("/success"), HTTP_GET, []() {
-    server.send(200, contentTypeHtml, successPage);
+  //
+  // Handle manual firmware updates at /update
+  //
+  server.on("/update", HTTP_GET, [](AsyncWebServerRequest *request){request->send(200,contentTypeHtml,updatePage);});
+  server.on("/update", HTTP_POST, [](AsyncWebServerRequest *request) {
+    // Check if the Update library encountered any errors.
+    bool shouldReboot = !Update.hasError();
+
+    // Create a response. The AJAX script is just looking for a successful HTTP status.
+    AsyncWebServerResponse *response = request->beginResponse((shouldReboot ? 200 : 500), "text/plain", (shouldReboot ? "OK" : "FAIL"));
+    response->addHeader("Connection", "close");
+    request->send(response);
+
+    // If successful, restart the ESP32 to boot into the new firmware
+    if (shouldReboot) restartTimer.once(0.5, []() { ESP.restart(); });
+  }, [](AsyncWebServerRequest *request, String filename, size_t index, uint8_t *data, size_t len, bool final) {
+    if (!index) {
+      // First chunk: Initialize the OTA Update
+      // UPDATE_SIZE_UNKNOWN tells the library to just accept chunks until 'final' is true
+      if (!Update.begin(UPDATE_SIZE_UNKNOWN, U_FLASH)) {
+        sendResponse(500,F("Update begin failed"),request);
+      }
+    }
+
+    // Write chunk data to the flash memory
+    if (!Update.hasError() && len) {
+      if (Update.write(data, len) != len) {
+        sendResponse(500,F("Update write failed"),request);
+      }
+    }
+
+    // Final chunk: Close the OTA process
+    if (final) {
+      if (!Update.end(true)) {
+        sendResponse(500,F("Update end failed"),request);
+      }
+    }
   });
 
   server.begin();     // Start the local web server
@@ -2942,11 +2954,8 @@ void setup(void) {
   if ((!crsCode[0] && !tubeId[0] && !busAtco[0]) || (!nrToken[0] && boardMode==MODE_RAIL)) {
     if (!apiKeys) showSetupKeysHelpScreen();
     else showSetupCrsHelpScreen();
-    // First time setup mode will exit with a reboot, so just loop here forever servicing web requests
-    while (true) {
-      yield();
-      server.handleClient();
-    }
+    // First time setup mode will exit with a reboot, so just loop here forever
+    while (true) { delay(10); }
   }
 
   configTime(0,0, ntpServer);   // Configure NTP server for setting the clock
@@ -2993,12 +3002,12 @@ void setup(void) {
 
   if (boardMode == MODE_RAIL) {
       progressBar(F("Initialising National Rail interface"),67);
-      altStationActive = setAlternateStation();  // Check & set the alternate station if appropriate
+      altStationActive = setAlternateStation(false);  // Check & set the alternate station if appropriate
       raildata = new raildataXmlClient();
       int res = raildata->init(wsdlHost, wsdlAPI, &raildataCallback);
       if (res != UPD_SUCCESS) {
         showWsdlFailureScreen();
-        while (true) { server.handleClient(); yield();}
+        while (true) {delay(1);}
       }
       progressBar(F("Initialising National Rail interface"),70);
       raildata->cleanFilter(platformFilter,cleanPlatformFilter,sizeof(platformFilter));
@@ -3018,6 +3027,19 @@ void setup(void) {
 
 void loop(void) {
 
+  if (touchEnabled) button.updateTouchState();
+
+  if (button.wasShortTapped()) {
+    if (isSleeping) {
+      // force awake
+      forcedAwake = true;
+    } else {
+      switchToNextMode();
+    }
+  } else if (button.wasLongTapped() && boardMode==MODE_RAIL && altStationEnabled && !isSleeping) {
+    softResetBoard(MODE_RAIL,true);
+  }
+
   // Check for firmware updates daily if enabled
   if (dailyUpdateCheck && millis()>fwUpdateCheckTimer) {
     fwUpdateCheckTimer = millis() + 3300000 + random(600000); // check again in 55 to 65 mins
@@ -3083,5 +3105,10 @@ void loop(void) {
       break;
   }
 
-  server.handleClient();
-}
+  if (softResetNeeded) {
+    softResetNeeded=false;
+    softResetBoard(MODE_LOADCONFIG,false);
+  }
+
+  if (manualUpdateCheck) doManualOtaCheck();
+}
\ No newline at end of file
